### PR TITLE
Move `HF.C.Utils` and some `Consensus.Utils` to `Data.SOP`

### PR DESCRIPTION
**Pull request by Jasagredo**
# Description

These modules defined combinators which are super general, much more than only Consensus' Hard Fork Combinator. Therefore I'm moving them to `Data.SOP` in preparation for moving them somewhere else (`sop-extras`?).

# Checklist

- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [x] The documentation has been properly updated
    - [x] New tests are added if needed and existing tests are updated
    - [x] Any changes affecting Consensus packages must have an entry in the appropriate `changelog.d` directory created using [`scriv`](https://github.com/input-output-hk/scriv). If in doubt, see the [Consensus release process](../ouroboros-consensus/docs/ReleaseProcess.md).
    - [x] Any changes in the Consensus API (every exposed function, type or module) that has changed its name, has been deleted, has been moved, or altered in some other significant way must leave behind a `DEPRECATED` warning that notifies downstream consumers.
    - [x] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [x] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [x] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [x] Reviewer requested

### CODE REVIEW

File: ouroboros-consensus/changelog.d/20230306_122628_javier.sagredo_move_hfc_utils.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+<!--
+### Patch
+
+- A bullet item for the Patch category.
+
+-->
+
+### Non-Breaking
+
+- Move `Util` modules that are related only to SOP to `Data.SOP`.
```
Comment (jorisdral) [inline ouroboros-consensus/changelog.d/20230306_122628_javier.sagredo_move_hfc_utils.md:16]: Maybe mention which modules are going to be deprecated in the future?

**Review by jorisdral — COMMENTED** [review]
Nice :smile:

File: ouroboros-consensus-test/test-consensus/Test/Consensus/HardFork/Combinator.hs [hunk 2]
```diff
@@ -22,7 +22,12 @@
 module Test.Consensus.HardFork.Combinator (tests) where
 
 import qualified Data.Map.Strict as Map
+import           Data.SOP.Counting
+import           Data.SOP.InPairs (RequiringBoth (..))
+import qualified Data.SOP.InPairs as InPairs
```
Comment (dnadales) [inline ouroboros-consensus-test/test-consensus/Test/Consensus/HardFork/Combinator.hs:27]: Probably not for this PR, but I wonder if it'd make sense to have one `Data.SOP` module that would re-export all these symbols.

**Review by dnadales — APPROVED** [review]

File: ouroboros-consensus/changelog.d/20230306_122628_javier.sagredo_move_hfc_utils.md (hunk 1 above)
Comment (Jasagredo) [reply ouroboros-consensus/changelog.d/20230306_122628_javier.sagredo_move_hfc_utils.md:16]: Done

File: ouroboros-consensus/src/Data/SOP/Strict.hs [hunk 3]
```diff
@@ -22,15 +22,21 @@ module Data.SOP.Strict (
...
+  , sequence_NS'
   , unZ
     -- * Injections
   , Injection
   , injections
+    -- * fn
+  , fn_5
     -- * Re-exports from @sop-core@
   , module Data.SOP
   , module Data.SOP.Constraint
```
[the remark this answers is not in the document]
Comment (Jasagredo) [reply ouroboros-consensus/src/Data/SOP/Strict.hs:42]: As long as they use the `NP` and `NS` from this module I think they will be strict still, right?

File: ouroboros-consensus/ouroboros-consensus.cabal [hunk 4]
```diff
@@ -173,8 +173,8 @@ library
                        Ouroboros.Consensus.Util.Orphans
                        Ouroboros.Consensus.Util.RedundantConstraints
                        Ouroboros.Consensus.Util.ResourceRegistry
-                       Ouroboros.Consensus.Util.Singletons
                        Ouroboros.Consensus.Util.STM
+                       Ouroboros.Consensus.Util.Singletons
```
[the remark this answers is not in the document]
Comment (Jasagredo) [reply ouroboros-consensus/ouroboros-consensus.cabal:177]: Let's pray we have `cabal format`

File: ouroboros-consensus/src/Data/SOP/Strict.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus/src/Data/SOP/Strict.hs:42]: My comment was a bit silly, you're right

Side note: I see that some of the other modules that now live under `Data.SOP` also provide strictness guarantees. SO, when/if we decide to make `Data.SOP.Strict` its own package, maybe we should see which of these modules should also be included

**Comment by Jasagredo** [comment]
bors cancel

**Pull request closed without merging**